core/src/main/java/org/apache/iceberg/FastAppend.java

```diff
@@ -83,9 +85,13 @@ protected Map<String, String> summary() {
 
   @Override
   public FastAppend appendFile(DataFile file) {
-    this.hasNewFiles = true;
-    newFiles.add(file);
-    summaryBuilder.addedFile(spec, file);
+    Preconditions.checkNotNull(file, "Invalid data file: null");
+    if (newFilePaths.add(file.path())) {
+      this.hasNewFiles = true;
```

Comment:
Nit: This flag could also be replaced by `newFilePaths.length() > 0`. We could also do this in a follow up PR.

Reply:
I agree but I didn't want to do any additional refactorings as part of this PR, unless people think I should include it. I can do a follow-up PR to refactor this